tests/basic_cxx_qt/cpp/main.cpp

```diff
@@ -5,225 +5,222 @@
-  CHECK(uint32Spy.count() == 1);
-  CHECK(types.getUint32() == 4);
-}
+  Q_OBJECT
+
+private Q_SLOTS:
+  // CXX-Qt allows basic interaction between C++ (with Qt) and Rust
+  void test_basic_interaction()
+  {
+    cxx_qt::my_object::MyObject obj;
+
+    obj.sayHi(QStringLiteral("Hello World!"), 32);
+
+    // Check that an invokable can be called and the return value is correct
+    const auto value = obj.doubleNumber(32);
+    qInfo() << "Double of 32 is:" << value;
+    QCOMPARE(value, 64);
+
+    // Track the signal count of numberChanged, stringChanged
+    QSignalSpy numberSpy(&obj, &cxx_qt::my_object::MyObject::numberChanged);
+    QSignalSpy stringSpy(&obj, &cxx_qt::my_object::MyObject::stringChanged);
+
+    // Check the number property
+    QCOMPARE(obj.getNumber(), 0);
+    QCOMPARE(numberSpy.count(), 0);
+    obj.setNumber(16);
+    QCOMPARE(numberSpy.count(), 0);
+    QCoreApplication::processEvents();
+    QCOMPARE(numberSpy.count(), 1);
+    QCOMPARE(obj.getNumber(), 16);
+
+    // Check the string property
+    QCOMPARE(obj.getString(), QString());
+    QCOMPARE(stringSpy.count(), 0);
+    obj.setString(QStringLiteral("Hello"));
+    QCOMPARE(stringSpy.count(), 0);
+    QCoreApplication::processEvents();
+    QCOMPARE(stringSpy.count(), 1);
+    QCOMPARE(obj.getString(), QStringLiteral("Hello"));
+
+    // Check the double number self
+    QCOMPARE(obj.getNumber(), 16);
+    QCOMPARE(numberSpy.count(), 1);
+    obj.doubleNumberSelf();
+    QCOMPARE(obj.getNumber(), 32);
+    QCOMPARE(numberSpy.count(), 1);
+    QCoreApplication::processEvents();
+    QCOMPARE(numberSpy.count(), 2);
+
+    qInfo() << "Number is:" << obj.getNumber()
+            << "String is:" << obj.getString();
+  }
+
+  // CXX-Qt allows basic interaction between C++ (with Qt) and Rust using Serde
+  void test_basic_interaction_serde()
+  {
+    cxx_qt::my_data::MyData data;
+
+    // Track the signal count of numberChanged, stringChanged
+    QSignalSpy numberSpy(&data, &cxx_qt::my_data::MyData::numberChanged);
+    QSignalSpy stringSpy(&data, &cxx_qt::my_data::MyData::stringChanged);
+
+    // Check that initial value of the deserialised data
+    QCOMPARE(data.getNumber(), 4);
+    QCOMPARE(data.getString(), QStringLiteral("Hello World!"));
+
+    // Check the number changed property
+    QCOMPARE(numberSpy.count(), 0);
+    data.setNumber(16);
+    QCOMPARE(numberSpy.count(), 0);
+    QCoreApplication::processEvents();
+    QCOMPARE(numberSpy.count(), 1);
+    QCOMPARE(data.getNumber(), 16);
+
+    // Check the string property
+    QCOMPARE(stringSpy.count(), 0);
+    data.setString(QStringLiteral("Hello"));
+    QCOMPARE(stringSpy.count(), 0);
+    QCoreApplication::processEvents();
+    QCOMPARE(stringSpy.count(), 1);
+    QCOMPARE(data.getString(), QStringLiteral("Hello"));
+
+    // Check that initial value of the deserialised data
+    QCOMPARE(data.asJsonStr(),
+             QStringLiteral("{\"number\":16,\"string\":\"Hello\"}"));
```

Comment:
For bonus points you could check if Qt's JSON parser parses this information correctly.

Reply:
Right, in the future we could make these tests more meaningful. I was thinking about sharing the Rust code here with qml_features. Then have a QML page that does something with that data, then have C++ tests here that can use the same Rust interface etc.